dockstore/dockstore
# Add update execution metrics endpoint

**Description**
Companion PRs:
- https://github.com/dockstore/dockstore-deploy/pull/742
- https://github.com/dockstore/dockstore-support/pull/479

This PR adds a PUT endpoint that allows executions that were previously submitted to Dockstore to be updated. Only optional fields can be updated to prevent important information, such as the execution ID and status, from being modified. This new endpoint does not allow aggregated metrics to be updated because they are deprecated (I vote for removing it, but maybe in 1.16).

Changes:
- <s>In order to efficiently find the execution to update, the S3 key must contain the execution ID. The file name in S3 was previously the current time in ms, but it's now the user-provided execution ID. This means that each file in the metrics S3 bucket contains only 1 execution, whereas it previously could contain multiple.</s> **UPDATE:** This PR change is reverted. The method of submitting executions via the POST endpoint is unchanged, i.e. each POST invocation sends the executions request body as one S3 file.
  - <s>This required a change in the POST metrics endpoint that submits metrics to S3. It now splits executions into individual files to send to S3. This appears to significantly slow down the Terra metrics ingestion, see https://github.com/dockstore/dockstore-support/pull/478 for the stats</s>
  - <s>The individual executions are still sent in the `ExecutionsRequestBody` object. The main reason is because this requires no change in the metrics aggregator.</s>
  - <s>Since the file name is the execution ID, which is user-provided, we check if the key exists in S3 before submitting it. We don't want the user to overwrite existing files in S3 using the POST endpoint.</s> **UPDATE:** Since the POST endpoint is unchanged, the file name is still the time of submission in milliseconds. It's no longer the execution ID as the PR previously proposed. We are **not** checking during the POST executions endpoint for duplicate execution IDs that already exist in S3 because it's inefficient when there are many files. We do check for duplicate IDs for executions in the same request body.
    - The responsibility is on the user to ensure that they don't provide duplicate IDs. This may change in future enhancements. If the user does provide duplicate IDs, the metrics aggregator in https://github.com/dockstore/dockstore-support/pull/479 will take the latest execution.
- Created a new `AggregatedExecution` class that extends from `Metrics` to represent aggregated metrics that users can submit otherwise `Metrics` would have the required `executionId` field that is not needed when our metrics aggregator posts metrics to our database. When we remove the deprecated aggregated executions, we would remove this class.
- There is a new GET execution endpoint that admins, curators, and platform partners can use to get an individual execution by execution ID. This may be used by partners if they need to see what they previously submitted or if they want to view the newly updated execution.


**Review Instructions**
In staging:
- Use the POST executions endpoint to submit metrics to Dockstore.
- Use the PUT executions endpoint to modify an execution that you submitted.
- Use the GET execution endpoint to verify that the execution was updated correctly.
- Verify that you can't update information that is required (view the schema on Swagger)

**Issue**
[SEAB-5943](https://ucsc-cgl.atlassian.net/browse/SEAB-5943)

**Security and Privacy**

If there are any concerns that require extra attention from the security team, highlight them here.

e.g. Does this change...
* Any user data we collect, or data location?
* Access control, authentication or authorization?
* Encryption features?

Please make sure that you've checked the following before submitting your pull request. Thanks!



[SEAB-5943]: https://ucsc-cgl.atlassian.net/browse/SEAB-5943?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
